### ISSUE TITLE
Ability to launch SourceJS via NodeJS Script

### ISSUE THREAD

**Issue by Alanoll** [sourcejs/Source#240]
Problem: If I have a NodeJS application, attempt to perform a **require(sourcejs)** so I can script and debug using VS my site's code (or any SourceJS plugins), I receive multiple errors.

I've been using the following setup for my SourceJS site development:

* Visual Studio 2015
* NodeJS Tools for Visual Studio

I created a "Javascript Project" in Visual Studio, so I can add the site to my group's build chain, and then also have a separate repository containing just my site's specs, and not the SourceJS source.

My NodeJS package.json includes just the single dependancy for sourcejs, so I can then perform an **npm install** to quick install my site for development/production.

**Comment by Alanoll** [comment]
I've got the issue fixed, and just following the contribution.md steps. If I screw up the process, let me know and I'll be sure not to do it again.

**Comment by operatino** [comment]
Hi,

Thanks for your input and patch suggestion. I will look deeper into the changes, to validate the proposed solution.

**Comment by operatino** [comment]
Have you tried spawning sub-process from your node script, instead of calling sourcejs server function directly? This will ensure that SourceJS will be isolated from your parent application, and also will enable it running on a separate CPU thread,

**Comment by Alanoll** [comment]
What if the *start* function was expanded to take in an **options** objects that could act as another way to configure the SourceJS server (like the command-line arguments)?

**Comment by operatino** [comment]
Hm, in case you don't have anything in you parent application, why do you even need to call SourceJS as a function? I haven't worked with iisnode, but 0.6.0-dev version is completely compatible with other PaaS like heroku.

Usually those remote runtime environments just call `npm start` or whatever you define as a starting commang + set some environment.

**Comment by operatino** [comment]
Here's also an example of the recommended (and only flow for 0.6.0) of starting SourceJS https://github.com/sourcejs/init/blob/npm/package.json. Could be that you've been looking into an old structure, where indeed the SourceJS code was a part of local project.

**Comment by operatino** [comment]
Thanks for the explanation, now I get what you are trying to achieve. It's looks like a valid reasoning, and I would be happy to make SourceJS more compatible with different IDE's and cloud hosting.

My point here is about checking how to make the solution simpler, with less potential code to maintain. Therefore few more questions:

* Does VS allows to run npm scripts, instead of direct `node script.js` definition? If yes, npm scripts allow combining different pre/post actions, which you can use for bootstrapping your application
* Does iisnode allow executing other scripts before the actual application start? Usually services like that allow defining custom starting flows, where you can

```
node bootstrap-the-application.js
node ./node_modules/sourcejs/app.js
```

> What if the start function was expanded to take in an options objects that could act as another way to configure the SourceJS server (like the command-line arguments)?

That's a good idea, and if we would go for exporting a set of programmable functions from SourceJS, we'll definitely need to have it.

**Comment by Alanoll** [comment]
> My point here is about checking how to make the solution simpler, with less potential code to maintain.
I can definitely get on board with that.

From the nodejstools docs [here](https://github.com/Microsoft/nodejstools/wiki/npm-Integration), looks like it's NPM integration is from a package management perspective, not from a project launch/running perspective.
Their pre/post looks to be using [grunt tasks](https://github.com/Microsoft/nodejstools/wiki/Pre-and-Post-Build-Actions), but from a project build perspective, where as I'm trying to achieve a more runtime update.

**iisnode** using the following to map the IIS handlers
```xml
<configuration>
<system.webServer>
<handlers>
<add name="iisnode" path="hello.js" verb="*" modules="iisnode" />
</handlers>    
</system.webServer>
</configuration>
```

Browsing the source code, it appears to just take the **path** and use it as the **node** command line argument.

**Comment by operatino** [comment]
Alright, looks like there's no other options (at least yet) rather then expose programmable API from SourceJS main file. Let's proceed with PR then.

**Comment by operatino** [comment]
PR merged #241
